akka-http-core/src/main/scala/akka/http/impl/util/One2OneBidiFlow.scala

```diff
@@ -87,8 +91,13 @@ private[http] object One2OneBidiFlow {
           } else throw new UnexpectedOutputException(element)
         }
         override def onUpstreamFinish(): Unit = {
-          if (pending == 0 && isClosed(inIn) && !innerFlowCancelled) complete(outOut)
-          else throw OutputTruncationException
+          if (pending > 0 || innerFlowCancelled) {
+            if (materializer.isInstanceOf[ActorMaterializer])
+              ActorMaterializer.downcast(materializer).logger.debug(
+                "Inner stream finished before inputs completed. " +
```

Comment:
unfortunately this is not reliable, this is exactly the issue. Eliminate this whole dance and only keep the check that the emitted elements must never be more than the consumed ones.